### PR TITLE
improve search results, search titles

### PR DESCRIPTION
**Pull request by mekarpeles**

### CODE REVIEW

**Comment by mekarpeles** [comment]
@bfalling this should have fixed our concerns:
- author images now showing where available
- rm publisher search -- should this go under MORE? @bfalling 
- adds advanced search feature to dropdown (clicking redirects to /advancedsearch)
- "title" copy instead of "books"
- singular instead of plural facets

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 1]
```diff
@@ -390,216 +390,251 @@ function passwordHide(){
...
+                '<li><a href="' + work.key + '"><img src="' + cover_url(work.cover_i) +
+                    '"/><span class="book-desc"><div class="book-title">' +
+                    work.title + '</div>by <span class="book-author">' +
+                    author_name + '</span></span></a></li>'
+            );
+        },
+        authors: function(author) {
+            console.log(author.photograph);
+            $('header .search-component ul.search-results').append(
+                '<li><a href="/authors/' + author.key + '"><img src="' + ("http://covers.openlibrary.org/a/olid/" + author.key + "-S.jpg") + '"/><span class="author-desc"><div class="author-name">' +
```
Comment (hornc) [inline openlibrary/plugins/openlibrary/js/ol.js]: Shoudln't we use https for image assets here?

**Comment by hornc** [comment]
👍 on all the copy changes from me

**Comment by bfalling** [comment]
@mekarpeles: No, just remove Search by Publisher for now. Thx!

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 2]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
+    var searchFacets = {
+        'title': 'books',
+        'author': 'authors',
+        'subject': 'subjects',
+        'advanced': 'advancedsearch'
+    };
+
+    var composeSearchUrl = function(q, json, limit) {
+        var facet_value = searchFacets[localStorage.getItem("facet")];
+        var url = ((facet_value === 'books')? '/search' : "/search/" + facet_value);
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Don't need surrounding parentheses. Add a space before `?`. Just use single-quotes for the string.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 3]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
+        if (json) {
+            url += '.json';
+        }
+        url += '?q=' + q;
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Do we need to url-encode `q`?

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 4]
```diff
@@ -1,35 +1,35 @@
...
-	var baseUrl = getUrl .protocol + "//" + getUrl.host +
+        var getUrl = window.location;
+        var baseUrl = getUrl .protocol + "//" + getUrl.host +
             "/" + getUrl.pathname.split('/')[1];
-	window.history.pushState({
+        window.history.pushState({
             "html": document.html,
             "pageTitle": document.title + " " + query,
-	}, "", baseUrl + "?id=" + query);
+        }, "", baseUrl + "?id=" + query);
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Do we need to url-encode `query`?

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 5]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
+    // Maps search facet label with value
+    var searchFacets = {
+        'title': 'books',
+        'author': 'authors',
+        'subject': 'subjects',
+        'advanced': 'advancedsearch'
+    };
+
+    var composeSearchUrl = function(q, json, limit) {
+        var facet_value = searchFacets[localStorage.getItem("facet")];
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Why the switch to double-quotes?

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 6]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
-	localStorage.setItem("facet", facet.toLowerCase());
-	$('header .search-facet-selector select').val(capitalize(localStorage.getItem("facet")))
-	$('header .search-facet-value').html(capitalize(localStorage.getItem("facet")));
-    	var url = ((localStorage.getItem("facet") === 'books')? '/search' : "/search/" + localStorage.getItem("facet"));
-	console.log(url);
-	$('.search-bar-input').attr("action", url);
+        var facet_key = facet.toLowerCase();
+
+        if (facet_key === 'advanced') {
+            localStorage.setItem("facet", '');
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Use single quotes for consistency? Do throughout.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 7]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
+            window.location.assign('/advancedsearch')
+            return;
+        }
+
+        localStorage.setItem("facet", facet_key);
+        $('header .search-facet-selector select').val(capitalize(facet_key))
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Missing final semi-colon here and below.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 8]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
-	    
-	    if (timeout) {
-		clearTimeout(timeout);
-	    } else if (execAsap) {
-		func.apply(obj, args);
-	    }
-	    timeout = setTimeout(delayed, threshold || 100);
-	};
+        var timeout;
+        return function debounced () {
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Remove space between function name and `()`. Heck, shouldn't we be using a JS linter so we don't have to think about things like that?

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 9]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
+            var obj = this, args = arguments;
+            function delayed () {
+                if (!execAsap)
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Let's be consistent (and a linter would help) about whether we surround one-line `if` bodies with braces or not. See above and below.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 10]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
-		if (search_query) {
-		    renderSearchResults(search_query);
-		}
-	    }
-	    enteredSearchMinimized = false;
-	    searchExpansionActivated = false;
-	    $('header .logo-component').removeClass('hidden');
-	    $('header .search-component').removeClass('search-component-expand');
-	}
+        if($(this).width() < 568){
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Extract magic number into constant with meaningful name, or at least add a comment.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 11]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
+            var author_name = work.author_name ? work.author_name[0] : '';
+            $('header .search-component ul.search-results').append(
+                '<li><a href="' + work.key + '"><img src="' + cover_url(work.cover_i) +
+                    '"/><span class="book-desc"><div class="book-title">' +
+                    work.title + '</div>by <span class="book-author">' +
+                    author_name + '</span></span></a></li>'
+            );
+        },
+        authors: function(author) {
+            console.log(author.photograph);
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: `console.log` is discouraged these days.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 12]
```diff
@@ -390,186 +390,251 @@ function passwordHide(){
...
-	if(!$(event.target).closest('header .hamburger-component .hamburger-button').length) {
-	    $('header .hamburger-dropdown-component').hide();
-	}
-	if(!$(event.target).closest('.dropclick').length) {
+    $('header .search-component .search-bar-input input').keyup(debounce(function(e) {
+        // ignore directional keys and enter for callback
+        if (![13,37,38,39,40].includes(e.keyCode)){
+            renderSearchResults($(this).val());
+        }
+    }, 500, false));
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Magic number. Here and below.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 13]
```diff
@@ -589,20 +654,20 @@ $().ready(function(){
     };
 
     $('header .dropdown-avatar').click(debounce(function() {
-	var dropdown = $('#main-account-dropdown');
-	if (dropdown.is(':visible') === true) {
-	    hideUser();
-	} else {
-	    dropdown.slideToggle(25);
+        var dropdown = $('#main-account-dropdown');
+        if (dropdown.is(':visible') === true) {
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: `=== true` unnecessary.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 14]
```diff
@@ -589,20 +654,20 @@ $().ready(function(){
...
+            hideUser();
+        } else {
+            dropdown.slideToggle(25);
             $(this).toggleClass('hover');
             var offUser = $(this);
             $(document).mouseup(function(offUser){
-		if($(offUser.target).parent("a").length==0){
+                if($(offUser.target).parent("a").length==0){
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/ol.js]: Spaces around equality operator. Change to `===`?

**Comment by mekarpeles** [comment]
We need to fix the bug where if a facet is selected which is not in the set of valid facets, that it defaults to "title". This case occurs when we switch branches or change facet names or add a facet.

**Comment by mekarpeles** [comment]
Merging at @bfalling's request and retroactively applying non-critical fixes

**Pull request merged**